akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -137,6 +144,129 @@ object GraphStageLogic {
   private object DoNothing extends (() ⇒ Unit) {
     def apply(): Unit = ()
   }
+
+  trait MinimalDeathWatch {
+    this: MinimalActorRef ⇒
+    import StageActorRef._
+    private[this] val _watchedBy = new AtomicReference[Set[ActorRef]](ActorCell.emptyActorRefSet)
+
+    def log: LoggingAdapter
+
+    override def isTerminated = _watchedBy.get() == StageTerminatedTombstone
+
+    //noinspection EmptyCheck
+    protected def sendTerminated(): Unit = {
+      val watchedBy = _watchedBy.getAndSet(StageTerminatedTombstone)
+      val wasAlreadyTerminated = watchedBy == StageTerminatedTombstone
```

Comment:
but you already aquired watchedBy in a local val, there is problem with reentrancy here. You are deriving a boolean from a value that already is on your stack.